### PR TITLE
vsr: sync uses correct view to go into recovering_head 

### LINKED ISSUE #1703: Crash: 8593423301425288917
<strong>Commit:</strong> ```37dd6f6ed3263be57d0284825937cf899b78f7ea``` <br><br><strong>Branches:</strong> main<br><br><strong>Duration to run seed in ReleaseSafe mode:</strong> 4s<br><br><strong>Stack Trace:</strong><br>
```
thread panic: reached unreachable code
src/vsr/replica.zig:0:76: in valid_hash_chain_between (simulator)
src/vsr/replica.zig:5801:49: in repair (simulator)
assert(self.valid_hash_chain_between(self.op_repair_min(), self.op));

src/vsr/replica.zig:2924:24: in tick (simulator)
self.repair();

src/testing/cluster.zig:410:37: in tick (simulator)
replica.tick();

src/simulator.zig:263:23: in main (simulator)
simulator.tick();

zig/lib/std/start.zig:574:37: in posixCallMainAndExit (simulator)
const result = root.main() catch |err| {

zig/lib/std/start.zig:243:5: in _start (simulator)
asm volatile (switch (native_arch) {

???:?:?: in ??? (???)
Unwind information for `???:0x1` was not available, trace may be incomplete

debug: exit with signal: 6. Indicates a crash bug.


```
<br><strong>Debug Logs:</strong><br><details><summary>Tail</summary><br>
```debug: Replaying seed 8593423301425288917 in Debug mode with full debug logging enabled...


SEED=8593423301425288917

replicas=3
standbys=3
clients=6
request_probability=42%
idle_on_probability=4%
idle_off_probability=10%
one_way_delay_mean=9 ticks
one_way_delay_min=1 ticks
packet_loss_probability=23%
path_maximum_capacity=17 messages
path_clog_duration_mean=239 ticks
path_clog_probability=1%
packet_replay_probability=33%
partition_mode=testing.packet_simulator.PartitionMode.uniform_partition
partition_symmetry=testing.packet_simulator.PartitionSymmetry.symmetric
partition_probability=2%
unpartition_probability=7%
partition_stability=144 ticks
unpartition_stability=19 ticks
read_latency_min=2
read_latency_mean=10
write_latency_min=0
write_latency_mean=62
read_fault_probability=7%
write_fault_probability=4%
crash_probability=0.00002%
crash_stability=984 ticks
restart_probability=0.0002%
restart_stability=641 ticks
 1/.1V0/__1/__1C0:__4Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?3/4Pp1/3Rq
 1/.1V0/__2/__2C0:__5Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?3/4Pp0/3Rq
 5|.1V0/__1/__2C0:__4Jo0/_2J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.1V0/__2/__2C0:__4Jo0/_2J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.1V0/__1/__2C0:__6Jo0/_6J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 2\.1V0/__1/__2C0:__4Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 2\.1V0/__2/__2C0:__4Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.1V0/__1/__2C0:__6Jo0/_6J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.1V0/__2/__2C0:__6Jo0/_6J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.1V0/__2/__2C0:__6Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .2V0/__1/__2C0:__6Jo0/_2J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .2V0/__2/__2C0:__6Jo0/_2J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0/ .3V0/__3/__3C0:__5Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?2/4Pp0/3Rq
 0/ .3V0/__4/__4C0:__5Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?1/4Pp0/3Rq
 2\.3V0/__3/__4C0:__5Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 2\.3V0/__4/__4C0:__5Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 1\.3V0/__3/__4C0:__5Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 1\.3V0/__4/__4C0:__5Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.3V0/__3/__4C0:__6Jo0/_2J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.3V0/__4/__4C0:__6Jo0/_2J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.3V0/__3/__4C0:__7Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.3V0/__4/__4C0:__7Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.3V0/__3/__4C0:__7Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.3V0/__4/__4C0:__7Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 2/.5V0/__5/__5C0:__8Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?3/4Pp2/3Rq
 2/.5V0/__6/__6C0:__9Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?3/4Pp1/3Rq
 2/.5V0/__7/__7C0:_10Jo0/_2J!0:_31Wo <__0:__0> v1:10Ga0G!0G?3/4Pp0/3Rq
 1\.5V0/__5/__7C0:__8Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 1\.5V0/__6/__7C0:__8Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 1\.5V0/__7/__7C0:__8Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.5V0/__5/__7C0:__8Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.5V0/__6/__7C0:__8Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/__5/__7C0:__8Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/__6/__7C0:__8Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 2/.5V0/__8/__8C0:_11Jo0/_2J!0:_31Wo <__0:__0> v1:10Ga0G!0G?3/4Pp0/3Rq
 1\.5V0/__8/__8C0:__8Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.5V0/__5/__8C0:_11Jo0/_6J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.5V0/__7/__8C0:__8Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.5V0/__8/__8C0:__8Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 2/.5V0/__9/__9C0:_12Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?3/4Pp0/3Rq
 1\.5V0/__9/__9C0:__9Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.5V0/__6/__9C0:_11Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.5V0/__7/__9C0:_11Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.5V0/__8/__9C0:_11Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.5V0/__9/__9C0:_11Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/__7/__9C0:_12Jo0/_4J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/__8/__9C0:_12Jo0/_4J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .5V0/__5/__9C0:_12Jo0/_4J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .5V0/__6/__9C0:_12Jo0/_4J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .5V0/__7/__9C0:_12Jo0/_4J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .5V0/__8/__9C0:_12Jo0/_4J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/__9/__9C0:_12Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 2/.5V0/_10/_10C0:_13Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?3/4Pp1/3Rq
 2/.5V0/_11/_11C0:_14Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?3/4Pp0/3Rq
 2/.5V0/_12/_12C0:_14Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?2/4Pp0/3Rq
 2/.5V0/_13/_13C0:_14Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?1/4Pp0/3Rq
 0\ .5V0/__9/_10C0:_14Jo0/_2J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .5V0/_10/_10C0:_14Jo0/_2J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .5V0/_11/_13C0:_15Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .5V0/_12/_13C0:_15Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .5V0/_13/_13C0:_15Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/_10/_13C0:_15Jo0/_4J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/_11/_13C0:_15Jo0/_4J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/_12/_13C0:_15Jo0/_4J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 2/.5V0/_14/_14C0:_15Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?1/4Pp0/3Rq
 2/.5V0/_15/_15C0:_15Jo0/_0J!0:_31Wo <__0:__0> v1:10Ga0G!0G?0/4Pp0/3Rq
 5|.5V0/__9/_13C0:_15Jo0/_7J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.5V0/_10/_13C0:_15Jo0/_7J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.5V0/_11/_13C0:_15Jo0/_7J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.5V0/_12/_13C0:_15Jo0/_7J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.5V0/_13/_13C0:_15Jo0/_7J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .5V0/_14/_15C0:_16Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 0\ .5V0/_15/_15C0:_16Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.5V0/_14/_15C0:_15Jo0/_6J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 5|.5V0/_15/_15C0:_15Jo0/_6J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.5V0/_10/_15C0:_19Jo0/_7J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.5V0/_11/_15C0:_19Jo0/_7J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.5V0/_12/_15C0:_19Jo0/_7J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.5V0/_13/_15C0:_19Jo0/_7J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/_13/_15C0:_15Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/_14/_15C0:_15Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 3|.5V0/_15/_15C0:_15Jo0/_1J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 1\.5V0/_10/_15C0:_15Jo0/_5J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 1\.5V0/_11/_15C0:_15Jo0/_5J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 1\.5V0/_12/_15C0:_15Jo0/_5J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 1\.5V0/_13/_15C0:_15Jo0/_5J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 1\.5V0/_14/_15C0:_15Jo0/_5J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 1\.5V0/_15/_15C0:_15Jo0/_5J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 2/.5V0/_16/_16C0:_19Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?3/4Pp2/3Rq
 2/.5V0/_17/_17C0:_20Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?3/4Pp1/3Rq
 2/.5V0/_18/_18C0:_21Jo0/_2J!0:_31Wo <__0:__0> v1:19Ga0G!0G?3/4Pp0/3Rq
 2/.5V0/_19/_19C0:_21Jo0/_2J!0:_31Wo <__0:__0> v1:19Ga0G!0G?2/4Pp0/3Rq
 0\ .5V0/_16/_19C0:_21Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 0\ .5V0/_17/_19C0:_21Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 0\ .5V0/_18/_19C0:_21Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 0\ .5V0/_19/_19C0:_21Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 4|.5V0/_14/_15C0:_19Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 4|.5V0/_15/_15C0:_19Jo0/_3J!0:_31Wo <__0:__0> v1:10Ga0G!0G?
 2/.5V0/_20/_20C0:_23Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?3/4Pp0/3Rq
 2/.5V0/_21/_21C0:_23Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?2/4Pp0/3Rq
 0\ .5V0/_20/_21C0:_21Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 0\ .5V0/_21/_21C0:_21Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 5|.5V0/_16/_21C0:_24Jo0/_8J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 5|.5V0/_17/_21C0:_24Jo0/_8J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 5|.5V0/_18/_21C0:_24Jo0/_8J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 5|.5V0/_19/_21C0:_24Jo0/_8J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1\.5V0/_16/_19C0:_21Jo0/_3J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1\.5V0/_17/_19C0:_21Jo0/_3J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1\.5V0/_18/_19C0:_21Jo0/_3J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 4|.5V0/_16/_21C0:_25Jo0/_6J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 4|.5V0/_17/_21C0:_25Jo0/_6J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 4|.5V0/_18/_21C0:_25Jo0/_6J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 4|.5V0/_19/_21C0:_25Jo0/_6J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 4|.5V0/_20/_21C0:_25Jo0/_6J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 5|.5V0/_20/_21C0:_24Jo0/_3J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 5|.5V0/_21/_21C0:_24Jo0/_3J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1/v7V0/_19/_21C0:_21Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1/v7V0/_20/_21C0:_21Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1/v7V0/_21/_21C0:_21Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 3|.8V0/_16/_21C0:_25Jo0/10J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 2/.8V0/_22/_22C0:_25Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?3/4Pp2/3Rq
 2/.8V0/_23/_23C0:_26Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?3/4Pp1/3Rq
 2/.8V0/_24/_24C0:_27Jo0/_2J!0:_31Wo <__0:__0> v1:19Ga0G!0G?3/4Pp0/3Rq
 0\ .8V0/_22/_24C0:_27Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 0\ .8V0/_23/_24C0:_27Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 0\ .8V0/_24/_24C0:_27Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1\.8V0/_22/_24C0:_27Jo0/_2J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1\.8V0/_23/_24C0:_27Jo0/_2J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1\.8V0/_24/_24C0:_27Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 2/.8V0/_25/_25C0:_27Jo0/_2J!0:_31Wo <__0:__0> v1:19Ga0G!0G?2/4Pp0/3Rq
 0\ .8V0/_25/_25C0:_27Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1\.8V0/_25/_25C0:_27Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 2/.8V0/_26/_26C0:_27Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?1/4Pp0/3Rq
 2/.8V0/_27/_27C0:_27Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?0/4Pp0/3Rq
 0\ .8V0/_26/_27C0:_27Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 0\ .8V0/_27/_27C0:_27Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 4|.7V0/_21/_21C0:_24Jo0/_3J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1\.8V0/_26/_27C0:_28Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 1\.8V0/_27/_27C0:_28Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 2 [ /.8V0/_27/_27C0:_30Jo0/_1J!0:_31Wo <__0:__0> v1:117Ga0G!0G?3/4Pp0/3Rq
 0 [ \ .8V0/_27/_27C0:_31Jo0/_1J!0:_31Wo <__0:__0> v1:117Ga0G!0G?
 1 [ \.8V0/_27/_27C0:_28Jo0/_0J!0:_31Wo <__0:__0> v1:117Ga0G!0G?
 3|.8V0/_17/_24C0:_25Jo0/_4J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 3|.8V0/_18/_24C0:_25Jo0/_4J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 3|.8V0/_19/_24C0:_25Jo0/_4J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 3|.8V0/_20/_24C0:_25Jo0/_4J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 3|.8V0/_21/_24C0:_25Jo0/_4J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 3|.8V0/_22/_24C0:_25Jo0/_4J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 3|.8V0/_23/_24C0:_25Jo0/_4J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 0 ] \ .8V23/_27/_27C0:_31Jo0/_0J!0:_31Wo <__0:__0> v1:117Ga0G!0G?
 3|.8V0/_24/_24C0:_25Jo0/_1J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 2 ] /.8V23/_27/_27C0:_31Jo0/_0J!0:_31Wo <__0:__0> v1:117Ga0G!0G?4/4Pp1/3Rq
 2/.8V23/_28/_28C0:_31Jo0/_0J!0:_31Wo <__0:__0> v1:117Ga0G!0G?3/4Pp1/3Rq
 2/.8V23/_29/_29C1:_32Jo0/_1J!0:_31Wo <__0:__0> v1:117Ga0G!0G?3/4Pp0/3Rq
 4|.8V0/_22/_27C0:_31Jo0/10J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 4|.8V0/_23/_27C0:_31Jo0/10J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 4|.8V0/_24/_29C0:_31Jo0/10J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 5 < ~.7V0/_21/_21C0:_24Jo0/_3J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 0\ .8V23/_28/_29C2:_33Jo0/_4J!0:_33Wo <__0:__0> v1:117Ga0G!0G?
 0\ .8V23/_29/_29C2:_33Jo0/_4J!0:_33Wo <__0:__0> v1:117Ga0G!0G?
 2/.8V23/_30/_30C2:_33Jo0/_0J!2:_33Wo <__0:__0> v1:118Ga0G!0G?3/4Pp0/3Rq
 1 ] \.8V23/_27/_29C0:_29Jo0/_0J!0:_31Wo <__0:__0> v1:117Ga0G!0G?
 1\.8V23/_28/_29C0:_29Jo0/_0J!0:_31Wo <__0:__0> v1:117Ga0G!0G?
 1\.8V23/_29/_29C0:_29Jo0/_0J!0:_31Wo <__0:__0> v1:117Ga0G!0G?
 3 < ~.8V0/_24/_30C0:_25Jo0/_0J!0:_31Wo <__0:__0> v1:19Ga0G!0G?
 5 > |.7V23/_23/_23C0:_24Jo0/_3J!0:_31Wo <__0:_27> v1:1nullGa0G!0G?
```
</details>

### PR DESCRIPTION
**Pull request by matklad**
It might be the case that op_checkpoint:

* is prepared  in view X
* is truncated in view X+1
* is committed in view X+2

When deciding whether to go into recovering_head state, the replica
currently considers prepared view (checkpoint.header.view), and not the
committed view.

Use the correct view by:

* Adding checkpoint's view into VSR State (but _not_ checkpoint state:
  different replicas might commit prepares at different views)
* Populating that view from the message that informed us about the
  checkpoint target

  * this requires some intricate logic on pings, to make sure they
    indeed propagate correct view for checkpoint --- a replica accepts a
    checkpoint before it transitions to its view, and it should
    subsequently correctly propagate this higher view.

    This works because checkpoint view is also durable.

Seed: 8593423301425288917
Closes: https://github.com/tigerbeetle/tigerbeetle/issues/1703

### CODE REVIEW

**Comment by matklad** [comment]
last commit fixes https://github.com/tigerbeetle/tigerbeetle/issues/1702 -- a semi related vopr seed that also trips on main

File: src/vsr/replica.zig [hunk 1]
```diff
@@ -2716,7 +2717,13 @@ pub fn ReplicaType(
             message.header.set_checksum_body(message.body());
             message.header.set_checksum();
 
-            assert(message.header.view <= self.view);
+            // Pings advertise checkpoints, and current checkpoint's view might be greater than
+            // the replica view.
+            if (message.header.view > self.view) {
```
Comment (sentientwaffle) [inline src/vsr/replica.zig:2722]: Should this be `> self.view_durable()`? We might have just called `transition_to_normal_from_recovering_head_status()` so our `view_durable` is not yet updated, but our `self.view` is updated.

**Review by sentientwaffle — COMMENTED** [review]
Should we update `sync_view` during `superblock.checkpoint()`? Then we would have the invariant that `sync_view >= checkpoint.header.view`.

File: src/vsr/replica.zig (hunk 1 above)
Comment (matklad) [reply src/vsr/replica.zig:2722]: I don't think so. I believe that

```
            if (message.header.view > self.view_durable()) {
                assert(self.superblock.working.vsr_state.sync_view >
                    self.superblock.working.vsr_state.view);
            }
```

would hold, but that's an almost tautological assertion. 

In contrast

```
            if (message.header.view > self.view_durable()) {
                assert(self.status == .recovering_head);
            }
```

would _not_ hold -- `transition_to_normal_from_recovering_head_status` changes status, but leaves `self.superblock.working.vsr_state` intact until durable update finishes. 


And

```
            if (message.header.view > self.view) {
                assert(self.status == .recovering_head);

```

Is the interesting case -- it ensures that we can't get out out `.recovering_head` state earlier than `self.superblock.working.vsr_state.view`

Comment (sentientwaffle) [reply src/vsr/replica.zig:2722]: Ahh I get it; thank you.

**Comment by matklad** [comment]
>Should we update sync_view during superblock.checkpoint()? Then we would have the invariant that sync_view >= checkpoint.header.view.

Excellent point! We should at least reset it to zero once the sync is done. But I think it's best not to update it, and to keep it scoped strictly to state sync, rather than mix in both sync and non-sync code paths in a single field.

File: src/vsr/superblock.zig [hunk 2]
```diff
@@ -958,6 +970,7 @@ pub fn SuperBlockType(comptime Storage: type) type {
             vsr_state.commit_max = update.commit_max;
```
Comment (sentientwaffle) [inline src/vsr/superblock.zig:972]: (A few lines above this) we can assert that `update.sync_view >= update.checkpoint.header.view`.

File: src/vsr/superblock.zig [hunk 3]
```diff
@@ -863,6 +873,7 @@ pub fn SuperBlockType(comptime Storage: type) type {
...
             vsr_state.sync_op_min = update.sync_op_min;
             vsr_state.sync_op_max = update.sync_op_max;
+            vsr_state.sync_view = update.sync_view;
```
Comment (sentientwaffle) [inline src/vsr/superblock.zig:876]: We could assert that `update.sync_view == 0 or update.sync_view == superblock.staging.vsr_state.sync_view`.

**Review by sentientwaffle — APPROVED** [review]

**Pull request merged**